README
ElasticSearch Import Utility (esimport)
=======================================

Tool exists to index content from a CSV in ElasticSearch.

Use
---

To index the contents of a CSV file (where the first row contains field
names) into an ElasticSearch server, you must supply the -s, -f, -i and
-t arguments. If the specified index or type does not exist within the
specified ElasticSearch server, it will be created when the module
executes. The following example will index data from data.file into
ElasticSearch at http://myserver:9200/myindex/mytype:

.. code:: Bash

    python -m esimport -s myserver:9200 -f /path/to/import/data.file -i myindex -t mytype

-  -s *server* (may either be a hostname:port or fully qualified, i.e.
   https://servername:port)
-  -f *filepath* (location of tab-delimited file to import data from)
-  -i *index*\ name\_ (name of the target index within ElasticSearch)
-  -t *type*\ name\_ (name of the target document type within
   ElasticSearch)

Further help available via the script:

.. code:: Bash

    python -m esimport --help

Additional Options
~~~~~~~~~~~~~~~~~~

Custom Delimiter
^^^^^^^^^^^^^^^^

The default delimiter for the operation is a ",", but you may specify different delimiters via the -d argument.

.. code:: Bash

    python -m esimport -s myserver:9200 -f /path/to/import/data.file -i myindex -t mytype -d '|'

- -d *delimiter* (the delimiter separating columns within the CSV)

Clear Existing Data First
^^^^^^^^^^^^^^^^^^^^^^^^^

When indexing data in ElasticSearch, you may optionally clear the
existing ElasticSearch type before data is added by including the -rm
argument.

.. code:: Bash

    python -m esimport -s myserver:9200 -f /path/to/import/data.file -i myindex -t mytype -rm

-  -rm (removes all documents of given type within the specified index
   on ElasticSearch)

Mapping
^^^^^^^

Mapping is the definitiion of the field names and value types of a
document indexed within ElasticSearch. Characteristics such as field
type, whether a field is searchable, and whether to include a timestamp
may all be defined in a map.

*Please note:* If the specified mapping does not match the field names
in the tab-delimited file, you should supply a field translation file
(described below).

A mapping file can be specified by adding the -m parameter to the
command:

.. code:: Bash

    python -m esimport -s myserver:9200 -f /path/to/import/data.file -i myindex -t mytype -m /path/to/mapping.json

-  -m *mapping*\ filepath\_ (location of JSON formatted mapping file for
   specified type)

Maps should be provided in JSON format, as seen below and on the
ElasticSearch website.

Sample mapping for a document type "tracks"
'''''''''''''''''''''''''''''''''''''''''''

.. code:: JSON

    {
        "tracks" : {
            "properties" : {
                "genre" : {
                    "type" : "string"
                },
                "album" : {
                    "type" : "string"
                },
                "ISRC" : {
                    "type" : "string"
                },
                "recordCompany" : {
                    "type" : "string"
                },
                "artist" : {
                    "type" : "string"
                },
                "songToken" : {
                    "type" : "integer"
                },
                "durationInSeconds" : {
                    "type" : "integer"
                },
                "title" : {
                    "type" : "string"
                }
            }
        }
    }

More details on mapping may be found in `ElasticSearch's mapping
reference <http://www.elasticsearch.org/guide/reference/mapping/>`__.

Field Translations
^^^^^^^^^^^^^^^^^^

If the original field names from the CSV file need to be altered or
filtered during the import, you may provide a field translations file.
The first row of this file should consist of the original field names,
separated by the specified delimiter; the second, the new names, again
separated by the specified delimeter.

*Please note:* Any original field names omitted from the first row, will
be omitted from the indexed data as well. This is a handy way to filter
the columns at time of indexing if that is necessary.

.. code:: Bash

    python -m esimport -s myserver:9200 -f /path/to/import/data.file -i myindex -t mytype -rm -m /path/to/mapping.json -n /path/to/field/name/translations.file

-  -n *field*\ translation\_filepath\_ (location of CSV field
   translation file for specified type)

Sample Field Name Translations File
'''''''''''''''''''''''''''''''''''

::

    Album_Category,Album_ID,ISRC,Record_Co,Song_Artist,Song_ID,Song_Secs,Song_Title
    genre,albumID,ISRC,recordCompany,artist,songID,duration,title

Basic Auth
^^^^^^^^^^

If login credentials are required, add the arguments -user and -pass

.. code:: Bash

    python -m esimport -s https://myserver.com -f /path/to/import/data.file -i myindex -t mytype -user exampleuser -pass examplepassword

Verify SSL
^^^^^^^^^^

If ElasticSearch requires SSL, the -nv setting can be used to bypass certificate verification if necessary.

.. code:: Bash

    python -m esimport -s https://myserver.com -f /path/to/import/data.file -i myindex -t mytype -user exampleuser -pass examplepassword -nv

Bulk Index Count
^^^^^^^^^^^^^^^^

You may specify the max number of records to index at time (defaults to 1000) by using the `-bc` argument.

.. code:: Bash

    python -m esimport -s https://myserver.com -f /path/to/import/data.file -i myindex -t mytype -user exampleuser -pass examplepassword -bc 500

Timeout
^^^^^^^

You may specify the timeout (defaults to 60) for communication with Elasticsearch using the `-T` argument..

.. code:: Bash

    python -m esimport -s https://myserver.com -f /path/to/import/data.file -i myindex -t mytype -user exampleuser -pass examplepassword -T 30

Dependencies
------------

-  Python libraries

   -  ``rawes``: `Github <https://github.com/humangeo/rawes>`__ \|
      `documentation <https://github.com/humangeo/rawes#rawes>`__

-  ElasticSearch



CHANGES
v0.1.1, 2013/06/09 -- Initial release.
v0.1.2, 2013/06/10 -- Updated documentation.
v0.1.3, 2013/06/12 -- Provided support for JSON parsing and an option for turning off SSL verification.
v0.1.4, 2013/06/14 -- Adding ability to specify size of bulk import
v0.1.5, 2013/06/18 -- Fixing two small bugs (-bc parameter should be optional and large CSV files weren't supported)
v0.1.6, 2013/06/25 -- Fixing bug where when fields specified in header keys file don't exist in CSV, an exception is thrown
v0.1.7, 2013/06/29 -- Fixes for exception that was thrown when field name is omitted, and when a timeout occurs communicating with ES
v0.1.8, 2013/06/30 -- Now displayes the total number of imported documents when complete
v0.1.9, 2013/10/03 -- Incorporating changes to set a default delimiter for the module


src/search-sync-agent/ssa.py
import sys
import os
import time

import argparse
import csv

import pyes
import requests


#
# TODO:
# - change key names to something more understandable and more appropriate
#   to the document store paradigm
# - allow user to create multiple types within an index
# - allow user to use a type name different from the index name
# - reorganize argparser so it only accepts certain arguments with the arguments
#   they're MEANT to work with.  it just ignores irrelevant args right now.
#


#
# Add the contents of a tab-delimited file (TDF) to an ElasticSearch database.
#
def add_tdf_to_elasticsearch(filename, es_server, index_name=None, \
                             delete_preexisting_index=False, mapping=None, \
                             es_basic_auth=None):
    remove_bom_from_utf8(filename)
    
    if index_name is None:
        filename_no_path = filename.split('\\')[-1]
        index_name = filename_no_path.split('.')[0].lower()
    
    conn = pyes.ES(es_server, basic_auth=es_basic_auth)
    index_is_ready = verify_es_index(index_name, conn, delete_preexisting_index)
    
    if mapping is not None:
        # assume for now that type name should be the same as index name
        mapping_result = apply_mapping(index_name, index_name, es_server, mapping, \
                                       es_basic_auth=es_basic_auth)

    sys.stdout.write("Adding TDF data to ElasticSearch...")
    sys.stdout.flush()

    start_time = time.time()
    if index_is_ready:
        doc_count = 0
        with open(filename, 'rb') as tdf_file:
            reader = csv.DictReader(tdf_file, delimiter='\t')

            for row in reader:
                # for now, index name matches the type
                conn.index(row, index_name, index_name)
                
                doc_count += 1
                if (doc_count % 1000) == 0:
                    sys.stdout.write(".")
                    sys.stdout.flush()

        conn.indices.refresh(index_name)

    elapsed_time = time.time() - start_time
    print "done!  Completed in " + str(elapsed_time) + " seconds."



# UTF-8 files generated on Windows machines generally start with a byte-order
# mark.  We need to remove the BOM so the DictReader doesn't use it as part
# of a column header.  If it doesn't have the BOM, this won't hurt anything.
def remove_bom_from_utf8(filename):
    with open(filename, 'rb') as f:
        original_contents = f.read()

    sys.stdout.write("Checking " + filename + " for BOM... ")
    sys.stdout.flush()

    start_time = time.time()

    decoded_contents = original_contents.decode('utf-8-sig').encode('utf-8')

    elapsed_time = time.time() - start_time
    print "done!  Completed in " + str(elapsed_time) + " seconds."
    

    # if the conversion changed anything, write the new version to the file
    if original_contents != decoded_contents:
        sys.stdout.write("BOM located and removed.  Writing new file... ")
        sys.stdout.flush()
        
        start_time = time.time()

        with open(filename, 'w') as f:
            f.write(decoded_contents)

        elapsed_time = time.time() - start_time
        print "done!  Completed in " + str(elapsed_time) + " seconds."
    else:
        print "No BOM found, so no file changes are required."



#
# Adds the contents of all files in a directory with the specified extension(s)
# to an ES database.
# Assumption: The original files are TDF format.
#
def add_tdfs_in_dir_to_elasticsearch(path, es_server, tdf_extensions=['.txt', '.tdf'], \
                             mapping_extensions=['.map'], delete_preexisting_indices=False,
                             es_basic_auth=None):
    add_all_start_time = time.time()
    
    directory = os.listdir(path)

    if not path.endswith("\\"):
        path += "\\"

    files_added_count = 0
    
    for f in directory:
        if f.endswith(tuple(tdf_extensions)):
            print f
            
            mapping_path = None
            if mapping_extensions is not None:
                index_name = f.split('.')[0]
                
                # look for a file that is the index name and ends with
                # the provided map extension
                for ext in mapping_extensions:
                    print path + index_name + ext
                    try:
                        path_to_check = path + index_name + ext
                        with open(path_to_check):
                            mapping_path = path_to_check
                            break
                    except IOError:
                        pass
            
            add_tdf_to_elasticsearch(path + f, es_server, \
                    delete_preexisting_index=delete_preexisting_indices, \
                    es_basic_auth=es_basic_auth, mapping=mapping_path)
            files_added_count += 1
            print ""
    
    add_all_elapsed_time = time.time() - add_all_start_time
    print "Adding data from " + str(files_added_count) + \
          " files to ElasticSearch took " + str(add_all_elapsed_time) + \
          " seconds."

    return files_added_count



#
# Verifies that a specified index exists and creates it if not.
# Optional behavior: delete existing index and recreate it.
#
def verify_es_index(index_name, conn, delete_existing_index=False):
    try:
        if delete_existing_index:
            conn.indices.delete_index_if_exists(index_name)

        conn.indices.create_index_if_missing(index_name)
        return True
    except:
        print "Unexpected error: ", sys.exc_info()

    return False



#
# Apply the mapping in the supplied file to a type under an index.
#
def apply_mapping(index_name, type_name, es_server, map_filename, es_basic_auth=None):
    # ensure the byte-order mark won't trip us up
    remove_bom_from_utf8(map_filename)
    
    with open(map_filename, 'rb') as f:
        mapping_data = f.read()
    
    # put the mapping to the index.  if ES server is '10.129.1.210:9200',
    # index_name is 'songindex', type_name is 'songtype', and
    # map_filename is \\skynyrd\Export\Dev\data\Song.map', this is equivalent to
    #     curl -XPUT 'http://10.129.1.210:9200/songindex/songtype/_mapping' -d @'\\skynyrd\Export\Dev\data\Song.map'
    put_mapping_address = 'http://' + es_server + '/' + index_name + \
                          '/' + type_name + '/_mapping'
    
    requests_basic_auth = None
    if es_basic_auth is not None:
        requests_basic_auth = requests.auth.HTTPBasicAuth(es_basic_auth['username'], \
                                                          es_basic_auth['password'])
    
    request = requests.put(put_mapping_address, data=mapping_data, \
                           auth=requests_basic_auth)
    return request



#
# Dev test method; not meant for inclusion in non-dev branches.
#
def test_es(es_server, es_basic_auth=None):
    conn = pyes.ES(es_server, basic_auth=es_basic_auth)

    q = pyes.query.MatchAllQuery()
    results = conn.search(q)
    
    num_results = len(results)
    max_to_print = 20
    for r in results[:max_to_print]:
        print r

    if num_results > max_to_print:
        print "Printed the first " + str(max_to_print) + " results."

    print str(num_results) + " results found!"
    print conn.get_indices()



#
# Set up the argparser in its own method, for clarity.
#
def set_up_argparser():
    # specify default extensions for TDF files
    default_tdf_extensions = ['.txt']
    default_tdf_extensions_string = ""

    for extension in default_tdf_extensions:
        default_tdf_extensions_string += extension
        if extension is not default_tdf_extensions_string[-1]:
            if len(default_tdf_extensions) > 1:
                default_tdf_extensions_string += "; "

    # specify default extensions for map files
    default_map_extensions = ['.txt']
    default_map_extensions_string = ""

    for extension in default_map_extensions:
        default_map_extensions_string += extension
        if extension is not default_map_extensions_string[-1]:
            if len(default_map_extensions) > 1:
                default_map_extensions_string += "; "

    parser = argparse.ArgumentParser(description='Process TDF files and add \
                                     them to an ElasticSearch database.')

    group = parser.add_mutually_exclusive_group(required=True)

    # mutually exclusive group
    group.add_argument('-f', '--files', nargs='+', \
                       help="TDF files that need to be added to ElasticSearch.")
    group.add_argument('-d', '--dirs', nargs='+', \
                       help="Directories containing TDF files that need to be \
                             added to ElasticSearch.  If -ext is not \
                             specified, only files ending in " + \
                             default_tdf_extensions_string + " will be processed.")
    group.add_argument('-test', '--run_test', \
                        action="store_true", default=False, \
                        help="Test query to verify presence of data.")
    group.add_argument('-indices', '--get_indices', action="store_true", \
                       default=False, help="Get summary of indices.")

    parser.add_argument('-tdf_ext', '--tdf_extensions', nargs='+', \
                        default=default_tdf_extensions, \
                        help="TDF file extensions to process when processing an \
                              entire directory.  If not specified, defaults \
                              to " + default_tdf_extensions_string)
    parser.add_argument('-map_ext', '--map_extensions', nargs='+', \
                        default=default_map_extensions, \
                        help="Map file extensions to process when processing an \
                              entire directory.  If not specified, defaults \
                              to " + default_map_extensions_string)
    parser.add_argument('-s', '--es_server', nargs=1, required=True, \
                        help="The ElasticSearch database's server and port.")
    parser.add_argument('-del', '--delete_preexisting_index', \
                        action="store_true", default=False, \
                        help="Before adding docs to an index, delete the index \
                              if it already exists.")

    parser.add_argument('-map', '--map_file_path', nargs=1, default=None, \
                        help="Path to file containing the desired mapping for \
                              this index.")
    
    parser.add_argument('-user', '--username', nargs=1, default=None, \
                        help="Username for ElasticSearch database.")
    parser.add_argument('-pass', '--password', nargs=1, default=None, \
                        help="Password for ElasticSearch database.")

    return parser.parse_args()



#
# The main method.
#
def main(argv):
    args = set_up_argparser()

    operation_start_time = time.time()

    # Construct user-friendly summary of operations run
    friendly_list_of_ops = []
    
    # args.es_server only allows a SINGLE argument at present (05/2013)
    # but is still parsed in by argparser as a single-item list.  Since
    # it's used in multiple places below, let's assign that single item
    # to a variable for convenience.
    es_server = args.es_server[0]
    
    # by setting this to None, we can pass the variable into the ES
    # connection method regardless of whether there's any basic auth data and 
    es_basic_auth = None
    
    if args.username is not None and args.password is not None:
        es_basic_auth = { 'username': args.username,
                          'password': args.password }

    if args.files is not None:
        for f in args.files:
            add_tdf_to_elasticsearch(f, es_server, \
                    delete_preexisting_index=args.delete_preexisting_index, \
                    mapping=args.map_file_path[0], es_basic_auth=es_basic_auth)
        
        friendly_list_of_ops.append("adding " + str(len(args.files)) + \
                                    " file(s) to the ElasticSearch database")

    if args.dirs is not None:
        files_converted = 0
        for d in args.dirs:
            files_converted += add_tdfs_in_dir_to_elasticsearch(d, \
                    es_server, tdf_extensions=args.tdf_extensions, \
                    delete_preexisting_indices=args.delete_preexisting_index, es_basic_auth=es_basic_auth)

        friendly_list_of_ops.append("adding " + str(files_converted) + \
                " files in " + str(len(args.dirs)) + " directories to the \
                 ElasticSearch database")

    if args.run_test == True:
        test_es(es_server, es_basic_auth=es_basic_auth)
        friendly_list_of_ops.append("running a test query on the ES database")

    if args.get_indices == True:
        print pyes.ES(es_server).get_indices()

    operation_elapsed_time = time.time() - operation_start_time

    if len(friendly_list_of_ops) > 0:
        operation_description = ""

        # construct description of what just happened
        for op in friendly_list_of_ops:
            # add separator between last op description and this one
            if friendly_list_of_ops[0] is op:    # first item
                pass
            elif friendly_list_of_ops[-1] is op: # last item
                operation_description += " and "
            else:
                operation_description += "; "
            
            operation_description += op.strip()

        # Would love to use .capitalize() here, but it makes everything
        # that is NOT the first char lowercase!
        operation_description = operation_description[:1].upper() + operation_description[1:]
        print operation_description + " took " + \
              str(operation_elapsed_time) + " seconds."



if __name__ == "__main__":
    main(sys.argv[1:])
